src/main/kotlin/com/manimdsl/frontend/Type.kt

```diff
@@ -1,34 +1,125 @@
 package com.manimdsl.frontend
 
-sealed class Type: ASTNode()
+import com.manimdsl.linearrepresentation.*
+import com.manimdsl.shapes.Rectangle
 
-sealed class PrimitiveType: Type()
-object NumberType: PrimitiveType() {
-    override fun toString(): String {
-        return "number"
-    }
-}
+// Types (to be used in symbol table also)
+sealed class Type : ASTNode()
+
+// Primitive / Data structure distinction requested by code generation
+sealed class PrimitiveType : Type()
+object NumberType : PrimitiveType()
 
-sealed class DataStructureType(open var internalType: Type, open val methods: HashMap<String, DataStructureMethod>): Type() {
+sealed class DataStructureType(
+    open var internalType: Type,
+    open val methods: HashMap<String, DataStructureMethod>
+) : Type() {
     abstract fun containsMethod(method: String): Boolean
     abstract fun getMethodByName(method: String): DataStructureMethod
+
+    /** Init methods are to return instructions needed to create MObject **/
+    abstract fun init(
+        identifier: String,
+        x: Int,
+        y: Int,
+        text: String,
+    ): Pair<List<ManimInstr>, MObject>
+
+    abstract fun initRelativeToObject(
+        identifier: String,
+        text: String,
+        moveRelativeTo: String
+    ): Pair<List<ManimInstr>, MObject>
+}
+
+interface DataStructureMethod {
+    fun animateMethod(
+        arguments: List<String> = emptyList(),
+        variableNameGenerator: NameGenerator = DummyNameGenerator,
+        animationFlags: AnimationFlags = AnimationFlags(),
```

Comment:
Animation stuff should not be in the frontend. Can we refactor this in a way such that there is greater separation between the Manim specific stuff and the frontend as from a software engineering perspective we should be able to easily target a new animation library if we wanted

Reply:
The only way I can think of doing this is having an exact copy on both frontend and backend with only backend having animateMethod only in backend, which means it's then bad software engineering to have to add the same class in 2 places? Let me know if you have any better ideas

Reply:
Or we have a big where clause somewhere and hard code all the animate methods inside that?

Reply:
I think the where clause is preferable to this

Reply:
Either when clause or method overloading but the problem here is it's a very natural question that will be raised

Reply:
Not really without redesigning the way we implement our data structures and their methods (which would be good to think about). Also for the same purpose we abstracted out the linear IR and flags, so to target a new library we can just change the python in the ManimInstr classes.

Reply:
Whenever you invoke animateMethod just call a function with a when clause passing in the type and it should be the same as far as I can tell?

Reply:
That way we can take the results from the ast as is and completely rewrite the backend if we wanted. The animateMethod is only really specific to the code generation so it's creating an unnecessary dependency